Tests/MongoSwiftTests/APMTests.swift

```diff
@@ -0,0 +1,240 @@
+#if compiler(>=5.5.2) && canImport(_Concurrency)
+import MongoSwift
+import Nimble
+import NIO
+import NIOConcurrencyHelpers
+import TestsCommon
+
+@available(macOS 10.15, *)
+
+private protocol StreamableEvent {
+    var type: EventType { get }
+}
+
+extension CommandEvent: StreamableEvent {}
+
+extension SDAMEvent: StreamableEvent {}
+
+final class APMTests: MongoSwiftTestCase {
+    func testClientFinishesCommandStreams() async throws {
+        let elg = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+        defer { elg.syncShutdownOrFail() }
+        let client = try MongoClient.makeTestClient(eventLoopGroup: elg)
+
+        // Task to create a commandEventStream and should return when the client is closed
+        let eventTask = Task { () -> Bool in
+            // let cmdStream = client.commandEventStream()
+            for try await _ in client.commandEventStream() {}
+            return true
+        }
+        try await client.db("admin").runCommand(["ping": 1])
+        try await client.close()
+
+        // Ensure the task returned eventually
+        _ = try await eventTask.value
+    }
+
+    func testCommandStreamHasCorrectEvents() async throws {
+        let elg = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+        defer { elg.syncShutdownOrFail() }
+        let client = try MongoClient.makeTestClient(eventLoopGroup: elg)
+        let commandStr: [String] = ["ping", "ping", "endSessions", "endSessions"]
+        let eventTypes: [EventType] = [
+            .commandStartedEvent,
+            .commandSucceededEvent,
+            .commandStartedEvent,
+            .commandSucceededEvent
+        ]
+
+        // Task to create a commandEventStream and should return when the client is closed
+        let eventTask = Task { () -> Int in
+            var i = 0
+            for try await event in client.commandEventStream() {
+                expect(event.commandName).to(equal(commandStr[i]))
+                expect(event.type).to(equal(eventTypes[i]))
+                i += 1
+            }
+            return i
+        }
+        try await client.db("admin").runCommand(["ping": 1])
+        try await client.close()
+
+        // Ensure the task returned with proper number of events
+        let numEvents = try await eventTask.value
+        expect(numEvents).to(equal(4))
+    }
+
+    func testCommandStreamBufferingPolicy() async throws {
+        let elg = MultiThreadedEventLoopGroup(numberOfThreads: 1)
```

Comment:
>I think its bc the client closes after we exit the scope of the closure so we cant actually cant reliably close the task and coerce it to return anything because we are no longer in the closure (happens in defer so we cant go back to older code)

Hmm, not sure I follow. When the client closes after the closure finishes, it should call `finish()` on the handler which should subsequently cause the task to exit. Is it possible you're awaiting the task while still within the closure? Note in my code sample above, I'm only awaiting on the value of the task outside of the closure provided to `withTestNamespace`.

>That is what I opted to implement here since the helper is desired, but it does use the "fire and forget" functionality that kaitlin warned against. I can also return from the self.withTestClient but am limited by scope here since we dont have access to i outside the task, so there isnt much more value than what I chose to do

Right, so the suggestion I had was to return the task from `withTestClient`, and then after that you can await on the task, which can return `i` or the events or what not.